pkg/controller/servicebinding/binder.go

```diff
@@ -165,23 +164,32 @@ func (b *binder) updateVolumes(volumes []interface{}) ([]interface{}, error) {
 	log := b.logger
 
 	log.Debug("Checking if binding volume is already defined...")
-	for _, v := range volumes {
-		volume, ok := v.(map[string]interface{})
-		if !ok {
-			return nil, err.New("type asserting volume to map of string to interface{} error")
+	volume := corev1.Volume{}
+	for k, v := range volumes {
+		jsonString, err := json.Marshal(v)
+		if err != nil {
+			return nil, errs.Wrapf(err, "could not marshal volume %s", v)
+		}
+		err = json.Unmarshal(jsonString, &volume)
+		if err != nil {
+			return nil, errs.Wrapf(err, "could not marshal volume %s", &volume)
 		}
-		if name == volume["name"] {
+
+		if volume.Name == name {
 			log.Debug("Volume is already defined!")
-			return volumes, nil
+			if volume.Secret.SecretName == b.sbr.Status.Secret {
+				return volumes, nil
+			}
+			volumes = append(volumes[:k], volumes[k+1:]...)
 		}
 	}
 
-	log.Debug("Appending new volume.", "Secret.Name", name)
+	log.Debug("Appending new volume.", "Secret.Name", b.sbr.Status.Secret)
 	bindVolume := &corev1.Volume{
 		Name: name,
 		VolumeSource: corev1.VolumeSource{
 			Secret: &corev1.SecretVolumeSource{
-				SecretName: name,
+				SecretName: b.sbr.Status.Secret,
```

Comment:
Depends on how the field is defined in the ServiceBindingStatus struct; I'd say it should not be a pointer, indicating that at creation time the value is an empty string and at some point it is set to a non empty string. 

At runtime, if the field is read and the value is empty it means reconciliation hasn't been processed yet. 

During runtime, this value should always be written to the correct value regardless the current state and the value used throughout the reconciliation process should be the expected one (or the computed one) instead of what's written in the ServiceBindingStatus. 

This also corroborates with transforming this whole thing to a very focused function, like the next incarnation of the createOrUpdate function that has been removed in this PR.